Script AMC recalculations

anagperal opened the pull request:
### :pushpin: References

-   **Issue:** Closes [AMC script for recalculations](https://app.clickup.com/t/8693ecv17)

### :memo: Implementation
- [x] Create script to allow recalculate all stored data of AMC product level and AMC substance level using current ATC data version every time in **datastore --> glass --> amc-recalculation** the value `recalculate` is true (using the organization units and periods added here)
- [x] Allow to build script for running it in background of DHIS2 server (as [metadata-synchronization scheduler](https://github.com/EyeSeeTea/metadata-synchronization))
- [x] Update README
- [x] Allow to create events with calculated data if current don't exist when flag `--calculate` is true 

### :video_camera: Screenshots/Screen capture
![Screenshot from 2024-01-19 10-38-40](https://github.com/EyeSeeTea/glass-dev/assets/49402898/95df2f1b-42e7-46ce-9326-82655e57966f)
![Screenshot from 2024-01-19 10-38-45](https://github.com/EyeSeeTea/glass-dev/assets/49402898/a25f3883-e38c-44bb-8607-eab95b7ca5f6)

![Screenshot from 2024-01-19 10-40-18](https://github.com/EyeSeeTea/glass-dev/assets/49402898/040eaf81-c15a-4310-baa4-7865a0a29c51)
![Screenshot from 2024-01-19 10-40-26](https://github.com/EyeSeeTea/glass-dev/assets/49402898/a296ff4d-2485-45f2-abf9-5942a1dd316d)


### :fire: Testing
1. Product Level data: in tracker capture check in Thailand in program AMC - Product Register a product with enrollment date in 2020 period, and check inside the values of the events of stage AMC - Raw Substance Consumption Calculated
2. Substance level data: In Capture in Thailand in AMC - Calculated Consumption Data program check events with report date in period 2020, open one and check values
3. Change current ATC version in datastore glass --> ATCs 
4. In datastore have in glass --> amc-recalculation:
```
{
  "date": "2024-01-19T09:54:04.091Z",
  "orgUnitsIds": [
    "vboedbUs1As"
  ],
  "periods": [
    "2020"
  ],
  "recalculate": false
}
``` 
And change recalculate to true
5. Start script: 
`yarn start-amc-recalculate --url "https://dev.eyeseetea.com/who-dev-238" --auth USERNAME:PASSWORD`
Flag to also create events with calculations if it does not exist: `yarn start-amc-recalculate --url "https://dev.eyeseetea.com/who-dev-238" --auth USERNAME:PASSWORD --calculate`
7. After the script ends, check again in Tracker Capture and in Capture if that events change ATC version to current one. In real life the ATC data of this version should change so also the values in the events will change

anagperal commented:
@ifoche before merging we need to agree on how often the script will look in datastore if it has to recalculate data, for now I have left it on every day

tokland commented on src/data/repositories/GlassATCDefaultRepository.ts:
```diff
@@ -64,14 +65,36 @@ export class GlassATCDefaultRepository implements GlassATCRepository {
+
+    disableRecalculations(): FutureData<void> {
+        return this.getRecalculateDataInfo().flatMap(recalculateDataInfo => {
+            if (recalculateDataInfo) {
```
[subjective] Not that this is wrong, but typically we use a guard here:  `if(!recalculateDateInfo) return Future.success(undefined);`, and we save an indentation level.

tokland commented on src/data/repositories/data-entry/AMCProductDataDefaultRepository.ts:
```diff
@@ -73,25 +74,21 @@ export class AMCProductDataDefaultRepository implements AMCProductDataRepository
-                TRACKER_IMPORT_STRATEGY_CREATE_AND_UPDATE
-            ).flatMap(response => {
+            return importApiTracker(this.api, { events: d2TrackerEvents }, importStrategy).flatMap(response => {
                 return Future.success(response);
```
I guess there was some code before, but as it stand, flatMap+success is a no-op, it can be removed.

tokland commented on src/data/repositories/data-entry/AMCSubstanceDataDefaultRepository.ts:
```diff
@@ -224,49 +326,153 @@ export class AMCSubstanceDataDefaultRepository implements AMCSubstanceDataReposi
+            return d2TrackerEvents;
+        } catch {
+            return [];
+        }
```
This block is copied verbatim above, except for the call, it can be abstracted.

tokland commented on src/domain/usecases/data-entry/amc/DisableAMCRecalculationsUseCase.ts:
```diff
@@ -0,0 +1,9 @@
+
+export class DisableAMCRecalculationsUseCase {
+    constructor(private atcRepository: GlassATCRepository) {}
+    public execute(): FutureData<void> {
```
[minor] blank line between methods

tokland commented on src/domain/usecases/data-entry/amc/RecalculateConsumptionDataProductLevelForAllUseCase.ts:
```diff
@@ -0,0 +1,303 @@
+import {
```
Without knowing the intricacies or history of the code, it seems that data repositories are huge, containing a lot of logic. Not saying we have to change this, but typically repositories should be logic-less, just get/store the data. The logic is in use cases and entities. Note more for the future than for that specific PR.

tokland commented on src/domain/usecases/data-entry/amc/utils/__tests__/calculationConsumptionSubstanceLevelData.spec.ts:
```diff
@@ -34,7 +34,7 @@ describe("Given calculate Consumption Substance Level Data function", () => {
             const orgUnitId = "vboedbUs1As";
             const rawSubstanceConsumptionData = givenRawSubstanceConsumptionDataByType(type);
-            const atcVersionsByKeys = atcVersionsByKeysData as ListGlassATCVersions;
+            const atcVersionsByKeys = atcVersionsByKeysData as unknown as ListGlassATCVersions;
```
These two forced casting may be introducing bugs. Review, it seems PATH should be defined as optional

anagperal replied:
Thanks for this because I've found that `ATCAlterationsData` type was not correctly typed and also in this json, ATC data was missing `PATH` (not used in these tests but it has to be in ATC data)

anagperal commented:
thanks @tokland for the review. I've implemented all the comments and also I've added into AMCSubstanceDataDefaultRepository and AMCProductDataDefaultRepository a TODO to move logic to use cases and entities instead of have more logic in repos.

ifoche reviewed (changes requested):
Please @anagperal if you could verify that:
- The scripts sets false in recalculate as it starts so any failure inside wouldn't produce an infinite loop of recalculations
- We tested it with 2023 and J01CE01 and it didn't work complaining about 
```
"ERROR - not calculate and go to next one - ddd_value_uploaded and ddd_unit_uploaded not found of: {"report_date":"2023-01-01T00:00:00.000","health_level_manual":1,"health_sector_manual":1,"salt_manual":"default","atc_manual":"J01CE01","packages_manual":11784541,"atc_version_manual":"2021","ddds_manual":3928180.333333334,"route_admin_manual":"P","data_status_manual":1,"tons_manual":1.964090166666667}
```

Could you please have a look? 
Thanks!

anagperal commented:
> * We tested it with 2023 and J01CE01 and it didn't work complaining about

@ifoche  the problem was:
- the atc_version_manual of this event is not ok. It should be a key like ATC-year-version and here is just 2021:
![Screenshot from 2024-01-30 17-19-23](https://github.com/EyeSeeTea/glass-dev/assets/49402898/d39f0860-4702-47e8-a92b-cf3981d99283)
- this is how recalculations in substance level work: the raw Substance Consumption Data events are taken to be used in the calculations. Once we have the new Calculated Consumption Data, then we match this data with current Calculated Consumption Data events and the values of these events are updated with the new ones. The problem with this testing is that we do not have current Calculated Consumption Data events to be updated with new calculations.
- For a good testing of substance level recalculations, there should be an event created in the program raw Substance Consumption Data and another events with the calculations in program Calculated Consumption Data. All data must match except Ddds autocalculated which is the data that is calculated with the algorithm

ifoche reviewed (changes requested):
thanks @anagperal . Some ideas from today's meeting to maybe complete the operation of it. Could you please implement them?
- If a substance do not have a calculated event, then produce an error in the logger to easily identify it
- Add a flag for the execution of the script enabling to not only recalculate but also calculate from scratch, producing the events if they don't exist
- Move the permanent information request calls that are immutable for all the calculation to an initial phase and then you don't retrieve them again. e.g. the periods and OUs for the calculation do not need to be fetched every single loop cycle.
- Add flag --debug to yarn start command (package.json) so we don't need to modify code to add it

anagperal commented:
thanks @ifoche , all comments have been implemented:

- [x] If a substance do not have a calculated event, then produce an error in the logger to easily identify it
-      logger.error(`Product level: these events could not be updated events=${eventIdsNoUpdated.join(",")}`);
-      logger.error(`Substance level: these events could not be updated events=${eventIdsNoUpdated.join(",")}`);

- [x] Add a flag for the execution of the script enabling to not only recalculate but also calculate from scratch, producing the events if they don't exist --> flag `--calculate`
- [x] Move the permanent information request calls that are immutable for all the calculation to an initial phase and then you don't retrieve them again. e.g. the periods and OUs for the calculation do not need to be fetched every single loop cycle. --> all info not related to OU or period now is taken once at the beginning
- [x] Add flag --debug to yarn start command (package.json) so we don't need to modify code to add it

ifoche reviewed (approved):
Thanks @anagperal . I'm finding still some little glitches but that I'm not going to made them block this PR, because very soon we will be editing again this logic as per my discussion today with the clients, so I will leave those improvements for that time. But just for the record:
- The flag --debug seems to be statically used in yarn start. What I was meaning is to allow yarn start to be run as yarn start --debug, parsing that debug flag and using it only when it is requested. Not a big deal anyways.
- On the other side I see some "undefined" on every loop that I don't know where they could be coming from:
![screenshots_1707495840_1724_09022024_1661x604](https://github.com/EyeSeeTea/glass-dev/assets/10931645/355d5f55-c95f-4418-9d74-3428e09033a7)

But as said. Let's treat those 2 in the near future when we will be modifying this script again
